src/NHibernate/Async/Engine/ForeignKeys.cs

```diff
@@ -0,0 +1,208 @@
+﻿//------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by AsyncGenerator.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+
+
+using NHibernate.Id;
+using NHibernate.Persister.Entity;
+using NHibernate.Proxy;
+using NHibernate.Type;
+
+namespace NHibernate.Engine
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	/// <content>
+	/// Contains generated async methods
+	/// </content>
+	public static partial class ForeignKeys
+	{
+
+		/// <content>
+		/// Contains generated async methods
+		/// </content>
+		public partial class Nullifier
+		{
+
+			/// <summary> 
+			/// Nullify all references to entities that have not yet 
+			/// been inserted in the database, where the foreign key
+			/// points toward that entity
+			/// </summary>
+			public async Task NullifyTransientReferencesAsync(object[] values, IType[] types, CancellationToken cancellationToken)
+			{
+				cancellationToken.ThrowIfCancellationRequested();
+				for (int i = 0; i < types.Length; i++)
+				{
+					values[i] = await (NullifyTransientReferencesAsync(values[i], types[i], cancellationToken)).ConfigureAwait(false);
+				}
+			}
+
+			/// <summary> 
+			/// Return null if the argument is an "unsaved" entity (ie. 
+			/// one with no existing database row), or the input argument 
+			/// otherwise. This is how Hibernate avoids foreign key constraint
+			/// violations.
+			/// </summary>
+			private async Task<object> NullifyTransientReferencesAsync(object value, IType type, CancellationToken cancellationToken)
+			{
+				cancellationToken.ThrowIfCancellationRequested();
+				if (value == null)
+				{
+					return null;
+				}
+				else if (type.IsEntityType)
+				{
+					EntityType entityType = (EntityType)type;
+					if (entityType.IsOneToOne)
+					{
+						return value;
+					}
+					else
+					{
+						string entityName = entityType.GetAssociatedEntityName();
+						return await (IsNullifiableAsync(entityName, value, cancellationToken)) .ConfigureAwait(false)? null : value;
+					}
+				}
+				else if (type.IsAnyType)
+				{
+					return await (IsNullifiableAsync(null, value, cancellationToken)) .ConfigureAwait(false)? null : value;
```

Comment:
This was fixed with [this](https://github.com/maca88/AsyncGenerator/commit/f0de6ad66ae672623e8854215379c940dbf807df) commit